### ISSUE TITLE
[Java] remove guava dependency

### ISSUE THREAD

**Issue by chaokunyang** [alipay/fury#1113]
## Is your feature request related to a problem? Please describe.
Fury use guava for generic reflection, cache. But guava is used so widely, it's very likely to version conflict with users code.

## Describe the solution you'd like
Remove guava dependency, implement all releated methods by fury self.

## Additional context
Shade guava is an option too, but introduce complexities  for fury GuavaSerializers and guava CVE

**Comment by nandakumar131** [comment]
@chaokunyang 
If you haven't started working on the guava removal part 2, I can take it up and post a PR.

**Comment by chaokunyang** [comment]
Hi @nandakumar131, thanks for the willingness to contribute to Fury. I haven't start part2 yet, you can take up this part work.

FYI, Currently Fury use guava TypeToken for generics inspection and use guava cache for some static data cache. These are the biggest dependency on guava

**Comment by nandakumar131** [comment]
Thanks @chaokunyang!
If the change becomes huge, I will split it into two PRs to make it easy to review.

**Comment by nandakumar131** [comment]
@chaokunyang  do we still want to support `GuavaCollectionSerializers`?

**Comment by chaokunyang** [comment]
> @chaokunyang do we still want to support `GuavaCollectionSerializers`?

I prefer to still support it, but may use a way without source api dependency. For example, we use methodhandle to get all method for invocation

**Comment by doom369** [comment]
Hello. It seems like all dependencies are removed, so you can now remove the guava from the required dependencies during the build? Currently, the fury jar is 2.3 mb and guava brings an additional 3 mb.

**Comment by Munoon** [comment]
This also helps in reducing JAR file size. Let me know if I can help you here.

**Comment by chaokunyang** [comment]
> This also helps in reducing JAR file size. Let me know if I can help you here.

That would be really great. Actually I am on the work of first release of Fury under asf and don't have enough time for this feature

**Comment by Munoon** [comment]
> @Munoon That would be really great. Actually I am on the work of first release of Fury under asf and don't have enough time for this feature

Can you please specify which exactly help you need? To complete @nandakumar131 PR or to do some further steps?

**Comment by chaokunyang** [comment]
> > @Munoon That would be really great. Actually I am on the work of first release of Fury under asf and don't have enough time for this feature
> 
> Can you please specify which exactly help you need? To complete @nandakumar131 PR or to do some further steps?

Here are the things which may need to be finished:
- Implement a `TypeRef` to replace `TypeToken` in guava, which is work that @nandakumar131 doing before.
- Remove guava Cache which Fury used for weak key and soft values map
- Remove guava `FinalizableWeakReference` which fury used to build a multi-key weak map
- Remove guava `ListeningExecutorService` which fury used to compile classes and register callback
- Remove guava `com.google.common.collect` usage in the src/main/java dir except `org.apache.fury.serializer.collection.GuavaCollectionSerializers`
- Make `org.apache.fury.serializer.collection.GuavaCollectionSerializers` load stub classes, so even guava jar are not in classpath, teh fury init won't throw exception, like we did in org.apache.fury.serializer.collection.ImmutableCollectionSerializers
- Add a script to check `com.google.` are only used in `GuavaCollectionSerializers` and test classes.

**Comment by Munoon** [comment]
OK, I'll separate those by different PRs starting from `TypeRef` implementation, which looks like the most hard one, so it'll take some time...

**Comment by chaokunyang** [comment]
> OK, I'll separate those by different PRs starting from `TypeRef` implementation, which looks like the most hard one, so it'll take some time...

Yep, the `TypeRef` is the most compilciated one, and the `TypeToken` in guava has some performance issue, so we did some optimization in `org.apache.fury.type.TypeUtils`. If we have our own implementation, such helper can be simplified a lot.

**Comment by nandakumar131** [comment]
@chaokunyang, sorry about the delay on #1273.
I will not be able to make any progress on the PR in next two weeks, I can take it up after that.

@Munoon, feel free to take over #1273, if you have started working on `TypeToken` changes.

**Comment by chrisco484** [comment]
I used Maven's dependency:tree goal and found that there's still quite a lot of Guava dependencies (using Fory 0.13.2).

I managed to reduce the set of dependencies to:

- com/google/guava - guava/32.1.2-jre
- com/google/guava - listenablefuture/9999.0-empty-to-avoid-conflict-with-guava
- com/google/guava - failureaccess/1.0.1

Things seem to work ok for my scenario at least - maybe if I trigger an error or some other condition I haven't hit yet it may give another ClassNotFound exception but for now it seems to only need the above 3 dependencies.

Is there another release available which can avoid some or all of these Guava dependencies?
